react/features/settings/components/web/video/VideoSettingsContent.js

```diff
@@ -0,0 +1,207 @@
+                jitsiTrack: null
+            })
+        };
+    }
+
+    /**
+     * Creates and updates the track data.
+     *
+     * @returns {void}
+     */
+    async _setTracks() {
+        const trackData = await createLocalVideoTracks(
+            this.props.videoDeviceIds,
+        );
+
+        this._disposeCurrentTracks();
+        this.setState({
+            trackData
+        });
+    }
+
+    /**
+     * Destroys all the tracks.
+     *
+     * @returns {void}
+     */
+    _disposeCurrentTracks() {
+        this.state.trackData.forEach(({ jitsiTrack }) => {
+            jitsiTrack && jitsiTrack.dispose();
+        });
+    }
+
+    /**
+     * Returns the click handler used when selecting the video preview.
+     *
+     * @param {string} deviceId - The id of the camera device.
+     * @returns {Function}
+     */
+    _onEntryClick(deviceId) {
+        return () => {
+            this.props.setVideoInputDevice(deviceId);
+            this.props.toggleVideoSettings();
+        };
+    }
+
+    /**
+     * Renders a preview entry.
+     *
+     * @param {Object} data - The track data.
+     * @param {number} index - The index of the entry.
+     * @returns {React$Node}
+     */
+    _renderPreviewEntry(data, index) {
+        const { error, jitsiTrack, deviceId } = data;
+        const { currentCameraDeviceId, t } = this.props;
+        const isSelected = deviceId === currentCameraDeviceId;
+        const key = `vp-${index}`;
+        const className = 'video-preview-entry';
+
+        if (error) {
+            return (
+                <div
+                    className = { className }
+                    key = { key }>
+                    <div className = 'video-preview-error'>{t(error)}</div>
+                </div>
+            );
+        }
+
+        const props: Object = {
+            className,
+            key
+        };
+
+        if (isSelected) {
+            props.className = `${className} video-preview-entry--selected`;
+        } else {
+            props.onClick = this._onEntryClick(deviceId);
+        }
+
+        return (
+            <div { ...props }>
+                <div className = 'video-preview-overlay' />
+                <Video
+                    className = { videoClassName }
+                    videoTrack = {{ jitsiTrack }} />
+            </div>
+        );
+    }
+
+    /**
+     * Implements React's {@link Component#componentDidMount}.
+     *
+     * @inheritdoc
+     */
+    componentDidMount() {
+        this._setTracks();
+    }
+
+    /**
+     * Implements React's {@link Component#componentWillUnmount}.
+     *
+     * @inheritdoc
+     */
+    componentWillUnmount() {
+        this._disposeCurrentTracks();
+    }
+
+    /**
+     * Implements React's {@link Component#componentDidUpdate}.
+     *
+     * @inheritdoc
+     */
+    componentDidUpdate(prevProps) {
+        if (this.props.videoDeviceIds.length !== prevProps.videoDeviceIds.length) {
```

Comment:
Yes the component is fully dependent on these videoDeviceIds.

Reply:
Can you explain why? What if it's just the device id that that has changed, but the length remained the same?

Reply:
Is it possible that only the device id changes without the actual number of devices to change? If the device gets a new id isn't the old one first removed? I'm not sure, I have no experience with this kind of behaviour, just asking...

Reply:
It may be the case right now, but you never  know, it's not  like things are set in stone. What worries me is that it relies on the order of events, specifically that all events will be always delivered correctly and none would be ever missed. I'm not sure if you can count on that with WebRTC, different OS versions and whatnot. Can you replace the length comparison with https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/redux/functions.js#L51 and we'll be done with it? If there's similar place for the audio device IDs please do that as well.

Reply:
Makes sense. I have updated it. This is the only place such a comparison is being made.